File: yyds-auto-mcp/src/tools/project.ts
/**
 * Script project management tools
 */

import { z } from "zod";
import { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { httpGetJson, httpPostJson } from "../client.js";

export function registerProjectTools(server: McpServer) {
  server.tool(
    "list_projects",
    "List all Python script projects on the device.",
    {},
    async () => {
      const projects = await httpGetJson<Array<Record<string, unknown>>>("/project/list");
      if (!Array.isArray(projects) || projects.length === 0) {
        return { content: [{ type: "text" as const, text: "No script projects found" }] };
      }
      const lines = projects.map((p) => `- ${p.name ?? p.projectName ?? "unknown"}`);
      return {
        content: [{ type: "text" as const, text: `Projects (${projects.length}):\n${lines.join("\n")}` }],
      };
    }
  );

  server.tool(
    "project_status",
    "Get the current script project's running status.",
    {},
    async () => {
      const res = await httpGetJson<{ running: boolean; project: string | null; error?: string }>("/project/status");
      if (res.error) throw new Error(res.error);
      return {
        content: [{ type: "text" as const, text: res.running ? `Running: ${res.project}` : "No project running" }],
      };
    }
  );

  server.tool(
    "start_project",
    "Start a script project by name.",
    {
      name: z.string().describe("Project name (from list_projects)"),
    },
    async ({ name }) => {
      const res = await httpGetJson<{ success?: boolean; error?: string }>(
        `/project/start?name=${encodeURIComponent(name)}`
      );
      if (res.error) throw new Error(res.error);
      return { content: [{ type: "text" as const, text: `Started project: ${name}` }] };
    }
  );

  server.tool(
    "stop_project",
    "Stop the currently running script project.",
    {},
    async () => {
      const res = await httpGetJson<{ success?: boolean; error?: string }>("/project/stop");
      if (res.error) throw new Error(res.error);
      return { content: [{ type: "text" as const, text: "Project stopped" }] };
    }
  );

  server.tool(
    "run_python_code",
    "Execute a Python code snippet on the device Python engine (interactive mode, runs in main process). Returns stdout/stderr output.",
    {
      code: z.string().describe("Python code string to execute"),
    },
    async ({ code }) => {
      const res = await httpPostJson<{
        success?: boolean;
        stdout?: string;
        stderr?: string;
        error?: string;
      }>("/engine/run-code", { code });
      if (res.error) throw new Error(res.error);
      const parts: string[] = [];
      if (res.stdout) parts.push(res.stdout);
      if (res.stderr) parts.push(`[stderr]\n${res.stderr}`);
      if (!parts.length) parts.push(res.success ? "OK (no output)" : "Failed (no output)");
      return {
        content: [{ type: "text" as const, text: parts.join("\n") }],
      };
    }
  );
}

File: yyds-android/app/src/main/assets/python-shims/yyds_auto/mixins/project.py
"""脚本项目管理 Mixin"""

from __future__ import annotations

from typing import TYPE_CHECKING, Dict, List, Optional

from ..types import ProjectInfo

if TYPE_CHECKING:
    from ..device import Device


class ProjectMixin:
    """脚本项目管理（yyds.auto 独有能力）"""

    _client: ...

    def list_projects(self) -> List[ProjectInfo]:
        """列出设备上的脚本项目"""
        result = self._client.get_json("/project/list")
        projects: List[ProjectInfo] = []
        if isinstance(result, list):
            for item in result:
                if isinstance(item, dict):
                    projects.append(ProjectInfo(
                        name=item.get("name", ""),
                        path=item.get("path", ""),
                        running=item.get("running", False),
                    ))
        return projects

    def start_project(self, name: str) -> None:
        """启动脚本项目

        Args:
            name: 项目名称
        """
        self._client.get_text("/project/start", params={"name": name})

    def stop_project(self) -> None:
        """停止当前运行的脚本项目"""
        self._client.get_text("/project/stop")

    def project_status(self) -> Dict:
        """获取当前项目运行状态"""
        return self._client.get_json("/project/status")

    def run_code(self, code: str, timeout: float = 30) -> str:
        """在设备上执行 Python 代码片段

        Args:
            code: Python 代码
            timeout: 超时秒数

        Returns:
            执行输出
        """
        result = self._client.post_text("/engine/run-code", {"code": code}, timeout=timeout)
        return result
